src/Promise.h

```diff
@@ -0,0 +1,36 @@
+#pragma once
+#include <QMutex>
+#include <QSemaphore>
+
+// Simple cancelable(!) promise because the machinery to do this
+// using std::future is still experimental. Also, QFuture would be
+// an option, except that it requires you to call finish() after
+// cancelling (and explicitly forbids finishing multiple times)
+// which makes using it in an not-necessarily-called callback
+// hard.
+template<typename T>
+struct Promise {
```

Comment:
I don't like the promise name here because, https://en.cppreference.com/w/cpp/thread/promise

I mean, I know it's a promise, but when I was reading the above code I thought we were using std::promise. Maybe Just rename it as OurPromise or CustomPromise?

Reply:
I'll rename it to CancelablePromise. That's its raison d'être.

Reply:
BreakablePromise? Too dramatic?

Reply:
```cpp
template<typename T> struct WeddingVow {};
```
You can break it, but AsyncService will leave you and you can never come back with new translation requests.

Reply:
I like that.